ocw/lib/gce.py

```diff
@@ -207,3 +209,112 @@ def cleanup_all(self):
                 if "warnings" in response:
                     for w in response["warnings"]:
                         self.log_warn(w["message"])
+        self.cleanup_vault_keys()
+
+    def cleanup_vault_keys(self):
+        self.log_info("## GCE Vault key cleaning - %s ##" % self.__project)
+        credentials = service_account.Credentials.from_service_account_info(
+            self.__credentials.getPrivateKeyData()
+        )
+        service = googleapiclient.discovery.build('iam', 'v1', credentials=credentials)
+        # now get service accounts
+        vault_emails = self.parser_service_account_list(service)
+        # for each vault account find the associated names with old keys
+        vault_keys = self.get_vault_name(service, vault_emails)
+
+        self.delete_vault_keys(service, vault_keys)
+
+    @filterAfterTime(hours=24)
+    def get_vault_name(self, service, vault_account_tuple):
+        ''' Returns the names of the Service Accounts filtered by date
+
+        Retieve the Service Accounts with the keys which have expired.
+        The expiration date is set to 24 hours.
+
+        Parameters
+        ----------
+        service : googleapiclient.discovery.Resource, required
+            The IAM Service Account instance
+        vault_account_tuple : iterable, required
+            An iterable object with strings of the Email address associated to the service accounts.
+        '''
+        keys = []
+        for email in vault_account_tuple:
+            service_resp = service.projects().serviceAccounts().keys().list(
+                name='projects/-/serviceAccounts/%s' % email).execute()
+            keys += [k for k in service_resp['keys']]
+        return keys
+
+    @filterService(name='vaultopenqa')
+    def parser_service_account_list(self, iam_service):
+        ''' Returns a list of the emails of the Service Accounts
+
+        The request is using the resource name of the project associated with
+        the service accounts. This returns a list of dict items, which are
+        also not vault related.
+        The objects returned in chunks which contain a `nextPageToken`
+        to the next page. The `parser_service_account_list` parses each page
+        and returns the full list. Subsequently, the list is filtered to return
+        only emails correlated with vault
+
+        Parameters
+        ----------
+        iam_service : googleapiclient.discovery.Resource, required
+            The IAM Service Account instance
+        '''
+        vault_emails = []
```

Comment:
what is "vault email" ?

Reply:
the take out from the `service_account['email']`

Reply:
service account is entity from google cloud `Credentials` , right ? how it is related to vault ?